kvdb-rocksdb/src/lib.rs

```diff
@@ -563,27 +546,30 @@ impl Database {
 						.expect("iterator params are valid; qed")
 				);
 
-				Some(DatabaseIterator {
-					iter: interleave_ordered(overlay_data, iter),
-					_marker: PhantomData,
-				})
+				Some(interleave_ordered(overlay_data, iter.map(|(k, v)| (k.into(), v.into()))))
 			},
 			None => None,
 		}
 	}
 
-	fn iter_from_prefix(&self, col: Option<u32>, prefix: &[u8]) -> Option<DatabaseIterator> {
+	fn iter_from_prefix<'a>(
+		&'a self,
+		col: Option<u32>,
+		prefix: &[u8]
+	) -> Option<impl Iterator<Item=(DBKey, DBValue)> + 'a>
+	{
 		match *self.db.read() {
 			Some(DBAndColumns { ref db, ref cfs }) => {
-				let iter = col.map_or_else(|| db.iterator_opt(IteratorMode::From(prefix, Direction::Forward), &self.read_opts),
-					|c| db.iterator_cf_opt(cfs[c as usize], IteratorMode::From(prefix, Direction::Forward), &self.read_opts)
-						.expect("iterator params are valid; qed"));
-
-				Some(DatabaseIterator {
-					iter: interleave_ordered(Vec::new(), iter),
-					_marker: PhantomData,
-				})
-			},
+				let iter = col.map_or_else(
+						|| db.iterator_opt(IteratorMode::From(prefix, Direction::Forward), &self.read_opts),
+						|c| db.iterator_cf_opt(
+							cfs[c as usize],
+							IteratorMode::From(prefix, Direction::Forward),
+							&self.read_opts,
+						).expect("iterator params are valid; qed"));
+
+				Some(interleave_ordered(Vec::new(), iter.map(|(k, v)| (k.into(), v.into()))))
```

Comment:
Since we're returning `impl Iterator` we don't need to interleave with an empty `Vec` anymore right? We should be able to just return `iter`.

Reply:
Probably ...
I am not sure I understand the reason for the interleaved-ordered in the master code. I guess it is to be as close as `iter` as possible?
I'll try removing it.